tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc

```diff
@@ -66,11 +69,74 @@ static int64_t multiply_dims(llvm::ArrayRef<int64_t> dims, int64_t res = 1) {
   return res;
 }
 
-static int64_t count_dynamic_dims(llvm::ArrayRef<int64_t> dims) {
-  int64_t count = 0;
-  for (auto dim : dims)
-    if (ShapedType::isDynamic(dim)) ++count;
-  return count;
+static OpFoldResult multiply_dims(ArrayRef<OpFoldResult> dims,
+                                  OpBuilder& builder, Location loc,
+                                  OpFoldResult init = nullptr) {
+  if (dims.size() == 0) {
+    return init ? init : OpFoldResult{builder.getIndexAttr(1)};
+  }
+
+  // Use the first value as the initial value
+  if (!init) {
+    init = dims.front();
+    dims = dims.drop_front();
+  }
+
+  auto multiply = [&](OpFoldResult x, OpFoldResult y) -> OpFoldResult {
+    auto val_x = getValueOrCreateConstantIndexOp(builder, loc, x);
+    auto val_y = getValueOrCreateConstantIndexOp(builder, loc, y);
+    return getAsOpFoldResult(
+        builder.createOrFold<arith::MulIOp>(loc, val_x, val_y));
+  };
+
+  return std::accumulate(dims.begin(), dims.end(), init, multiply);
+}
+
+static int64_t count_dynamic_dims(ArrayRef<int64_t> dims) {
+  return llvm::count_if(dims, ShapedType::isDynamic);
+}
+
+// Try to extract the known integer value for the given OpFoldResult
+// with the intention of interpreting it as a tensor dimension.
+// When the value is not statically known, produce ShapedType::kDynamic.
+static int64_t as_static_dimension(OpFoldResult value) {
```

Comment:
Same with this helper.

Reply:
I've chosen to rename this function to something a little more understandable. I'd like to keep this function just because it is used both directly and as the functor to `map_to_vector` across multiple functions.